Csv pipeline

lindavin opened the pull request:
# Description

This PR implements the CSV Pipeline for OED. This allows OED users to upload CSV files for meter or readings data.

Fixes #370 and fixes #398. 

I would like to thank @LouisLu705 for helping me at various points in this PR.

## Type of change

- [ ] Note merging this changes the node modules
- [ ] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request

## Limitations
- The page for uploading CSVs have inputs for all the documented parameters for the CSV Pipeline. At the moment, several of them are not functioning; however, I believe these parameters are implemented in the recent PR #629. These are:
     - Cumulative
     - Cumulative Reset
     - Cumulative Reset Start
     - Cumulative Reset End
     - Length
     - Length Variation
     - Update
- The error messages and success messages appear as pop up notification. These disappear after a period of time and it may be better to have them appear within a static error box rather than as temporary pop ups.
- The error messages from validation the parameters are generated by server depending on the type of error that has occurred. These are more useful for cURL requests, but the messages are full length strings that are concatenated together. The limitation here is that they are not translated and only in English.
- Additionally, the messages from the server are not 'ids' found in `data.json` and therefore not translatable by react-intl. 
- Many new strings are added to `data.json` that now need translation.
- At the end, the form is quite long and if more parameters are added in the future, it may get long. I think it could be better to be a multistage form.

## Screenshots
### New CSV Page
![image](https://user-images.githubusercontent.com/55451477/116824009-d5a71980-ab55-11eb-8cd3-6afef380a4bf.png)

lindavin committed:
Begin csv shape

lindavin committed:
Update csv package

lindavin committed:
Begin csv pipeline

lindavin committed:
Fix typo

lindavin committed:
Begin csv route

lindavin committed:
Merge branch 'csvPipeline' of https://github.com/lindavin/OED into csvPipeline

lindavin committed:
Merge branch 'development' into csvPipeline

lindavin committed:
Add missing variable

lindavin committed:
Add stub switch, formatting, and add params to cb

lindavin committed:
Take success and failure request functions from obvius and add todos

lindavin committed:
Add some query validation

lindavin committed:
Add meter creation if not found

lindavin committed:
Add csv route to server and first test to the csv route

lindavin committed:
Add test for uploading meters and fix errors found in test

Start the readme file and fix issues in the csv route found in testing.

lindavin committed:
Add test for uploading a reading for a specific meter and fix errors

We also removed the header from the csv file.

lindavin committed:
Add test for uploading a reading for a non-existent meter and fix errors

lindavin committed:
Remove file generated

lindavin committed:
Add tests for unimplemented features

lindavin committed:
Allow uploading meter via csv

lindavin committed:
Large rewrite

We rewrite the way that errors are handled and saved. We temporarily disable the uploading meters for a rewrite and remove a mistaken comment in the test.

lindavin committed:
Move validate password to its own middleware

lindavin committed:
Rewrite error message

lindavin committed:
Rename exports

lindavin committed:
Separate out function that saves csv

lindavin committed:
Add default param and remove unneeded validation

lindavin committed:
Insert missing new keyword

lindavin committed:
Fix logging errors

lindavin committed:
Fix final errors introduced by new error system

lindavin committed:
Set default values for params

lindavin committed:
Set default params when not provided

lindavin committed:
Split meter uploads and readings uploads into separate functions

lindavin committed:
Convert indentation

lindavin committed:
Modify meter uploads so that multiple meters can be uploaded in one file and fix error in the mapRow method

The mapRow method in the Meter class forgot to include the identifer field.

lindavin committed:
Add parameter to exclude csv headerrow

lindavin committed:
Remove unnecessary print statements

lindavin committed:
Convert pipeline to accept csv.gzip files rather than csv files

lindavin committed:
Create meters according to Meter class

lindavin committed:
Remove completed todo

lindavin committed:
Move validate password to its own file

lindavin committed:
Split mode into routes

lindavin committed:
Add password validation before file processing

lindavin committed:
Remove unused files

lindavin committed:
Remove unused test file

lindavin committed:
Add comments and change quotes

lindavin committed:
Merge branch 'development' into csvPipeline

lindavin committed:
Patch incomplete merge

lindavin committed:
Remove todo and add comment

lindavin committed:
Merge branch 'development' into csvPipeline

lindavin committed:
Remove note and todo

lindavin committed:
Rename test file

lindavin committed:
Add validation for checking that duplications is an integer

lindavin committed:
Add defaults constants so that they are easier to change

lindavin committed:
Add validation via jsonschema

lindavin committed:
Remove redundant check

lindavin committed:
Check that at least csv file has been submitted

lindavin committed:
Remove redundant check

lindavin committed:
Upade validation with formatting and validation function

lindavin committed:
Update with headers and formatting

lindavin committed:
Add gzip param

lindavin committed:
Update error messages.

lindavin committed:
Create form to submit readings as well as change body params to be camelCase and update README

lindavin committed:
Adds tabs as well as a form to submit meter data

lindavin committed:
Resolves some formatting issues

lindavin committed:
Change paths

lindavin committed:
Add try/catch

lindavin committed:
Fix error with shift

Credit to @LuisLu705 for the catch

lindavin committed:
Fix linter errors and add status codes into Error

lindavin committed:
Merge branch 'development' into csvPipeline

lindavin committed:
Fix field ordering issue found by @LouisLu705

lindavin committed:
Fix linter and typescript errors

lindavin committed:
Merge branch 'development' into csvPipeline

lindavin committed:
Improve UI styling

lindavin committed:
Move csv to its own page

lindavin committed:
Remove unused component

lindavin committed:
Update tests to pass

lindavin committed:
Create temp directory if not exists

lindavin committed:
Better organize code and add coments

lindavin committed:
Add comments

lindavin committed:
Modify uploading meters to handle new columns

lindavin committed:
Add remaining parameters

lindavin committed:
Trigger refresh readings

lindavin committed:
Fix tests to zip in memory

lindavin committed:
Fix authentication for curl or web requests

lindavin committed:
Move CSV components to own directory

lindavin committed:
Move CSV container to own directory

lindavin committed:
Add translations

lindavin committed:
Fix curl fix server error

lindavin committed:
Exclude test files from header check, thanks to help from @LouisLu705

lindavin committed:
Add success notifications

lindavin committed:
Update paths for mocha tests

lindavin committed:
Fix file not found error

lindavin committed:
Fix previous calls to loadCsvInput

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -0,0 +1,228 @@
+import translate from '../../utils/translate';
+import { FormattedMessage } from 'react-intl';
+
+/** A range of values, inclusive lower bound and exclusive upper bound. */
```
Might be nice to use JSDOC and state it returns an array of these values.

lindavin replied:
Thanks done.

huss commented on src/client/app/containers/RouteContainer.ts:
```diff
@@ -9,18 +9,22 @@ import { State } from '../types/redux/state';
 function mapStateToProps(state: State) {
 	const currentUser = state.currentUser.profile;
 	let loggedInAsAdmin = false;
+	let role: UserRole | null = null;
```
spacing issue not from current PR but lets fix it (line below)

lindavin replied:
Thanks done.

huss commented on src/client/app/containers/csv/UploadCSVContainer.tsx:
```diff
@@ -0,0 +1,283 @@
+		},
+		uploadReadingsPreferences: {
+			createMeter: false,
+			cumulative: false,
```
In src/server/sql/meter/create_meters_table.sql (lastest one in development, not yet in this PR), it has default values for these preferences for the DB. I think it might be good to make these the same.

lindavin replied:
Thanks, done.

huss commented on src/client/app/containers/csv/UploadCSVContainer.tsx:
```diff
@@ -0,0 +1,283 @@
+			...previousState,
+			uploadReadingsPreferences: {
+				...previousState.uploadReadingsPreferences,
+				duplications: value
```
Should this be setting the duplications state of something with time sort?

lindavin replied:
You're right, thanks for the catch. I have fixed it.

huss commented on src/client/app/utils/api/UploadCSVApi.ts:
```diff
@@ -0,0 +1,35 @@
+		for (const [preference, value] of Object.entries(uploadPreferences)){
+			formData.append(preference, value.toString());
+		}
+		formData.append('csvfile', readingsFile); // It is important for the server than the file is attached last.
```
than -> that

lindavin replied:
Thanks, done.

huss commented on src/server/middleware/validatePassword.js:
```diff
@@ -0,0 +1,16 @@
+const { CSVPipelineError } = require('../services/csvPipeline/CustomErrors');
+
+// STUB, TODO: Validate password
+async function validatePassword(password){
```
formatting/spacing

This does not seem to be used anywhere and it does not seem to be something that should be here unless a real password is used. I suspect it was from testing and should be removed but not certain.

lindavin replied:
Thanks for the catch. I have deleted it.

huss commented on src/server/routes/csv.js:
```diff
@@ -0,0 +1,139 @@
+					// email and password to be supplied for curl requests.
+					const { email, password } = request.body;
+					const verifiedUser = await verifyCredentials(email, password, true);
+					if(verifiedUser){
```
formatting

lindavin replied:
Thanks, done.

huss commented on src/server/routes/csv.js:
```diff
@@ -0,0 +1,139 @@
+		if (req.body.gzip === 'true') {
+			fileBuffer = zlib.gunzipSync(fileBuffer);
+		}
+		const filepath = await saveCsv(fileBuffer, 'meters');
```
Should the file be named meters or readings?

lindavin replied:
It should be named readings. Thanks for the catch.

huss commented on src/server/services/csvPipeline/saveCsv.js:
```diff
@@ -0,0 +1,23 @@
+const crypto = require('crypto');
+const { CSVPipelineError } = require('./CustomErrors');
+const fs = require('fs').promises;
+const streamBuffers = require('stream-buffers');
```
This and the next imports do not seem to be used.

lindavin replied:
Thanks I have deleted them.

huss commented on src/server/services/csvPipeline/uploadMeters.js:
```diff
@@ -0,0 +1,49 @@
+ * @param {express.Request} req 
+ * @param {express.Response} res 
+ * @param {filepath} filepath Path to meters csv file.
+ * @param conn 
```
Should say what conn is

lindavin replied:
Thanks, done.

huss commented on src/server/services/csvPipeline/uploadMeters.js:
```diff
@@ -0,0 +1,49 @@
+ */
+async function uploadMeters(req, res, filepath, conn) {
+	try {
+		// const columns = Object.keys(new Meter()).slice(1); // used for the shape of the csv
```
Should this commented out code remain?

lindavin replied:
Thanks, I have removed it.

huss commented on src/server/services/csvPipeline/validateCsvUploadParams.js:
```diff
@@ -0,0 +1,181 @@
+		cumulative: 'false',
+		cumulativeReset: 'false',
+		cumulativeResetStart: '0:00:00',
+		cumulativeResetEnd: '23:59:59',
```
The DB create_meters_table.sql has slightly different defaults. We should use those since they were the final ones decided. Sorry for the change while you were doing this work.
cumulative_reset_end TIME DEFAULT '23:59:59.999999',
I don't know if these defaults makes sense here but we do need to figure out how they related to DB defaults:
    reading_length TIME DEFAULT '00:00:00',
    reading_variation TIME DEFAULT '23:59:59.999999',

lindavin replied:
Thanks, done.

lindavin replied:
As for the DB defaults of reading_length and reading_variation, I suppose these would correspond to the parameters length and lengthVariation. However, I have not change the default values for length and lengthVariation since the devDocs has specified certain defaults.

lindavin replied:
I'm not so sure what to do about reading_length and reading_variation. Wouldn't they appear in the meters csv file and so we do not have to specify them as parameters here?

huss replied:
I agree that the meter will often have the values. Thinking about that, it would probably be best to load the meter values into the form but I think that is something for the future given it would be an update once the meter name is known. Given the DB will either have a set value or the default value, we should let it be used unless a specific value is given in the form/curl request. Thus, I have changed my mind and now think we should not have default values in the form for these values so the pipeline will use the DB defaults unless a new value is given. Thus, src/client/app/utils/csvUploadDefaults.ts would not set a default for cumulative for those 4 values. I think the other defaults are fine. Unfortunately we don't have duplication in the DB but we may change how it works so that is probably okay at this point (making default 1). Does this seem okay as a change?

lindavin replied:
This change sounds okay. To clarify, does this mean that we won't have the cumulative values in the request anymore as it would be known in the database?

huss replied:
Sorry for the confusion. What I think might be best is to allow the user to input a value (to override what is in the DB) but not to give default values in the form. Thus, I expect the pipeline to use the DB  values if nothing is given in the form but use any provided values as a override. Another consequence of this change is that the form and curl requests will act similarly for these values.

lindavin replied:
Thanks for clarifying. This is interesting because that means that the cumulative options are optional. Now that I think about it, checkboxes may give the illusion that not ticking means that option is ignored, but they are implicitly saying `false`. I will think more on whether or not I should use other button types. Maybe a radio button for true/false or a select button to _really_ ignore the option and rely on server/db defaults.

lindavin replied:
This is interesting because it means that the cumulative checkboxes need to represent three values (true,false,meter value) rather than just two values (true,false). I might have to change the button type to more appropriately serve this purpose. I will think more about it.

lindavin replied:
I decided not to change the check boxes. It would easier to just tick the check boxes depending on the value of the meter served from the DB, which would be a future task. The default value is undefined for parameters whose default should be read from the DB. I elaborate in `/src/server/services/csvPipeline/validateCsvUploadParams.js` in the recent commit 22a2252. Otherwise, done.

huss commented on src/server/services/csvPipeline/validateCsvUploadParams.js:
```diff
@@ -0,0 +1,181 @@
+		length: '',
+		lengthVariation: '',
+		refreshReadings: 'false',
+		timeSort: 'increasing'
```
I assume this relates to TimeSortTypes but since it is js and not ts cannot be used. If so, a comment that they must (?) be the same would be useful.

lindavin replied:
Okay done.

huss commented on src/server/test/web/csvPipelineTest.js:
```diff
@@ -0,0 +1,107 @@
+			return meter;
+		}));
+		expect(dbMeters).to.deep.equal(csvMeters);
+		expect((await Meter.getAll(conn)).length).to.equal(3);
```
Might be better to check the length before the deep.equal to avoid longer error message and the length is a more basic test.

lindavin replied:
Thanks for the suggestion! Done.

huss commented on src/server/test/web/csvPipelineTest.js:
```diff
@@ -0,0 +1,107 @@
+const metersPathWithHeader = `${__dirname}/csvPipeline/sampleMetersWithHeader.csv`;
+const metersWithHeaderBuffer = fs.readFileSync(metersPathWithHeader);
+const zippedMetersWithHeaderBuffer = zlib.gzipSync(metersWithHeaderBuffer);
+
```
The tests are fine. Would it be good to add basic test for unzipped file (meter and readings)?

lindavin replied:
Okay done.

huss committed:
Merge branch 'development' of https://github.com/OpenEnergyDashboard/OED into pr/lindavin/633

huss committed:
modify test data

- The identifier now needs to be unique.
- added sample readings with header.
- .gz sampleReadings now same as unzipped one.

huss reviewed (changes requested):
I have done testing of the new code and it works fine. I did not test all the options on upload since the new pipeline is needed for many of them. This is good work. A few overall comments:

- The /tmp/uploads/ files are not deleted. I cannot recall if I said to keep them but I don't think it is a good idea. We don't want to fill up the disk with CSV files. Thus, they should be deleted.
- I'm not positive but I think files may also be left in that directory when a failure occurs. For example, I saw “sampleReadings.csv-1620914667189.octet-stream”. I have not looked more into this.
- I merged with development since there was a conflict. A few changes were made to the test CSV files and pushed back to this PR.

lindavin committed:
Resolve most of the comments

lindavin committed:
Resolve comments for tests

lindavin committed:
Make defaults consistent

We made the defaults on the form UI be consisten with the defaults
specified in the database and with those specified in
/src/server/validateCsvUploadParams.

lindavin committed:
Clean up files after upload

lindavin commented:
Thank you for the review. I have addressed and resolved all of the comments. I left one of the threads unresolved because although I believe I fixed it, I'm not sure what to do about readings_length and reading_variation. I also don't know how to address the CodeQL alerts/errors.

lindavin committed:
Remove extraneous code and fix comments

huss commented:
If I am uploading a non-zip file, then I get message such as:

> web_1       | POST /api/csv/meters 200 234.403 ms - 55
> web_1       | [INFO@2021-05-17T17:03:19.621Z] Successfully deleted the uploaded file /usr/src/app/src/server/tmp/uploads/csvPipeline/2021-05-17T17:03:19.526Z-1d76-sampleMeters.csv.
> web_1       | 
> web_1       | [ERROR@2021-05-17T17:03:19.627Z] Failed to remove the file /usr/src/app/src/server/tmp/uploads/csvPipeline/2021-05-17T17:03:19.526Z-1d76-sampleMeters.csv.
> web_1       | Stacktrace: 
> web_1       | Error: ENOENT: no such file or directory, unlink '/usr/src/app/src/server/tmp/uploads/csvPipeline/2021-05-17T17:03:19.526Z-1d76-sampleMeters.csv'

It happens for readings and meters; also for form or curl. For zip files, it seems to delete two different files and works fine. I have not tried to track this down any more than running the code.

huss commented:
I'm having an issues (I feel as if it was not happening before the recent changes but not certain) where the refresh of readings does not seem to be happening. Not sure what is up.

huss commented:
I have reviewed and tested the recent changes. All seems good. Just a few comments already posted.

lindavin committed:
Fix redudant file removal error

lindavin committed:
Correct mistake on refreshReadings

lindavin commented:
I have fixed the issue regarding the following comment:

> If I am uploading a non-zip file, then I get message such as:
> 
> > web_1       | POST /api/csv/meters 200 234.403 ms - 55
> > web_1       | [INFO@2021-05-17T17:03:19.621Z] Successfully deleted the uploaded file /usr/src/app/src/server/tmp/uploads/csvPipeline/2021-05-17T17:03:19.526Z-1d76-sampleMeters.csv.
> > web_1       |
> > web_1       | [ERROR@2021-05-17T17:03:19.627Z] Failed to remove the file /usr/src/app/src/server/tmp/uploads/csvPipeline/2021-05-17T17:03:19.526Z-1d76-sampleMeters.csv.
> > web_1       | Stacktrace:
> > web_1       | Error: ENOENT: no such file or directory, unlink '/usr/src/app/src/server/tmp/uploads/csvPipeline/2021-05-17T17:03:19.526Z-1d76-sampleMeters.csv'
> 
> It happens for readings and meters; also for form or curl. For zip files, it seems to delete two different files and works fine. I have not tried to track this down any more than running the code.

Thanks for the catch.

As for the refresh of readings, I noticed a mistake in an `if` statement and have corrected it in the recent commit afe5035. There is probably a better way to do branching with parameters than to check against `'true'`, but maybe we can leave that as a future task. This _should_ fix any issues with refreshReadings and call it when specified. 

Moreover, I might be be misunderstanding the purpose of refreshReadings, but I think I can still view uploaded data for previous days even when I do not specify `refreshReadings=true` in my curl request.

lindavin committed:
Remove form default for cum reset start and end values

lindavin committed:
Read default values from DB for cumulative and cumulativeReset

lindavin commented:
I have addressed the comments since the second review. I left some TODOs in 22a2252, regarding how we should read `cumulativeResetStart` and `cumulativeResetEnd` from the DB and how the cumulative values should be served from the database and populate the form.

lindavin committed:
Remove leftover console log

huss committed:
add new params for meter creation

The code was using the name for the timezone due to missing parameters.

huss committed:
fix refreshReadings on CSV upload

By requiring a file that automatically ran the function, the
refreshReadingViews only executed the first time. It has been changed to:
- Not automatically execute the function but the code now executes it.
- The command to run in package.json was updated to match.
- Use a variable to control refresh to parallel gps.
- remove tslint disable in refresh since no longer outputs to console.
This was not a new change but cleanup.

huss commented:
I pushed fixes for a couple of small items. Everything seems to now be working fine and all comments were addressed. The CodeQL issues will be addressed outside this PR. Thanks to @lindavin for the good work and getting everything ready.
